<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment</b>
    
Hello, my name is Sveta Nosova and I am going to review this project. 

Before we start, I want to pay your attention to the color marking:
    
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
Great solutions and ideas that can and should be used in the future are in green comments.   
</div>    
    
    
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>

Yellow color indicates what should be optimized. This is not necessary, but it will be great if you make changes to this project.
</div>      
    
    
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>

Issues that need to be corrected to get right results are indicated in red comments. Note that the project cannot be accepted until these issues are resolved.
</div>    

<hr>
    
**Please, use some color other than those listed to highlight answers to my comments.**
I would also ask you **not to change, move or delete my comments** so that it would be easier for me to navigate during the next review.
    
In addition, my comments are defined as headings. 
They can mess up the content, however, they are convenient, since you can immediately go to them. I will remove the headings from my comments in the next review. 
   
    
    
<hr>
    
<font color='dodgerblue'>**A few words about the project:**</font> you did a very good job, everything is clear and neat, there are analysis and intermediate conclusions, the code is pretty much clear and easy to understand. I still have some questions, I have written them in my comments. I've also left there some recommendations for improving the project.
    
    
I will wait for the project for a second review :)

    
    
📌 Here are some tips and tricks that may help you with Markdown cells:    
<hr style="border-top: 3px solid purple; "></hr>

You can leave comments using this code inside a Markdown cell:
    
    
    <div class="alert alert-info">
    <h2> Student's comment</h2>

    Your text here. 
    </div>

    
    
    <font color='red'> This code is used to change text color. </font>     

<font color='red'> It will look like this. </font> 
    
If you don't want your comments to be headings, replace **h2** with **b** or just add `<a class="tocSkip">` after the phrase *Student's comment*.


You can find out how to **format text** in a Markdown cell or how to **add links** [here](https://sqlbak.com/blog/jupyter-notebook-markdown-cheatsheet) и [and here](https://medium.com/analytics-vidhya/the-ultimate-markdown-guide-for-jupyter-notebook-d5e5abf728fd).
</div>

<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment 2</b>


Thank you for sending the corrected version of the project. You've done a great job, I really appreciate it. The headings from the old comments are removed. I've also left a few comments titled as **Reviewer's comment 2**. Check them out, I've left some code examples for you. 
    

Your project has passed code review. Congratulations 😊
    

    
Take care and good luck! 😊   
</div>

# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan.

[In this notebook you're provided with hints and brief instructions and thinking prompts. Don't ignore them as they are designed to equip you with the structure for the project and will help you analyze what you're doing on a deeper level. Before submitting your project, make sure you remove all hints and descriptions provided to you. Instead, make this report look as if you're sending it to your teammates to demonstrate your findings - they shouldn't know you had some external help from us! To help you out, we've placed the hints you should remove in square brackets.]

[Before you dive into analyzing your data, explain the purposes of the project and hypotheses you're going to test.]

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
There's an introduction, which is good. It is important to write an introductory part, because it gives an idea about the content of the project.
</div>

## Open the data file and have a look at the general information. 

importing the libraries and loading the data:

In [106]:
# Loading all the libraries
import pandas as pd

# Load the data
try:
    data = pd.read_csv('credit_scoring_eng.csv')
except:
    data = pd.read_csv('/datasets/credit_scoring_eng.csv')

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
It's very good that you use `try-except`. Although it is usually used for catching some specific exceptions, it can be used here as well. Another good practice is to add `else` and `finally`. Thus, `try` block allows us to handle some error. The `except` block catches the exceptions. `else` does something if there's no error. `finally` is to be executed in any case, whether there is an exception or not (it is useful when you definitely need to do something, for example, to close a file). 
    
    
Look at this code snippet:    
</div>

```python

try:
    # Try to do smth here.
except Exception as e:
    # Handle an instance of SomeException.
    # So we deal with the expected error.
else:
    # Do smth if there is no exception.
finally:
    # Do something no matter what.
    
```

<div class="alert alert-info">
<b> Student's comment</b>

Thanks! good to know.
</div>

<font color='red'></font>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

😉
</div>

## Task 1. Data exploration

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan

Let's explore our data:

In [107]:
# Let's see how many rows and columns our dataset has
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [108]:
# let's print the first N rows
data.head(15)

children  days_employed  dob_years            education  education_id  \
0          1      -8437.673         42    bachelor's degree             0   
1          1      -4024.804         36  secondary education             1   
2          0      -5623.423         33  Secondary Education             1   
3          3      -4124.747         32  secondary education             1   
4          0     340266.072         53  secondary education             1   
5          0       -926.186         27    bachelor's degree             0   
6          0      -2879.202         43    bachelor's degree             0   
7          0       -152.780         50  SECONDARY EDUCATION             1   
8          2      -6929.865         35    BACHELOR'S DEGREE             0   
9          0      -2188.756         41  secondary education             1   
10         2      -4171.484         36    bachelor's degree             0   
11         0       -792.702         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0      -1846.642         54         some college             2   
14         0      -1844.956         56    bachelor's degree             0   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate

In [109]:
data.describe(include='all')

children  days_employed  dob_years            education  education_id  \
count  21525.000      19351.000  21525.000                21525     21525.000   
unique       NaN            NaN        NaN                   15           NaN   
top          NaN            NaN        NaN  secondary education           NaN   
freq         NaN            NaN        NaN                13750           NaN   
mean       0.539      63046.498     43.293                  NaN         0.817   
std        1.382     140827.312     12.575                  NaN         0.548   
min       -1.000     -18388.950      0.000                  NaN         0.000   
25%        0.000      -2747.424     33.000                  NaN         1.000   
50%        0.000      -1203.370     42.000                  NaN         1.000   
75%        1.000       -291.096     53.000                  NaN         1.000   
max       20.000     401755.400     75.000                  NaN         4.000   

       family_status  family_status_id gender income_type      debt  \
count          21525         21525.000  21525       21525 21525.000   
unique             5               NaN      3           8       NaN   
top          married               NaN      F    employee       NaN   
freq           12380               NaN  14236       11119       NaN   
mean             NaN             0.973    NaN         NaN     0.081   
std              NaN             1.420    NaN         NaN     0.273   
min              NaN             0.000    NaN         NaN     0.000   
25%              NaN             0.000    NaN         NaN     0.000   
50%              NaN             0.000    NaN         NaN     0.000   
75%              NaN             1.000    NaN         NaN     0.000   
max              NaN             4.000    NaN         NaN     1.000   

        total_income           purpose  
count      19351.000             21525  
unique           NaN                38  
top              NaN  wedding ceremony  
freq             NaN               797  
mean       26787.568               NaN  
std        16475.451               NaN  
min         3306.762               NaN  
25%        16488.505               NaN  
50%        23202.870               NaN  
75%        32549.611               NaN  
max       362496.645               NaN

We can see that the columns names defined correctly.
While checking the data sample we can learn a few things:
1. The column days_employed has negative values and missing values, we need to check why.
2. The column education has inconsistent values, there are lower and upper case letters.
3. The column children min is negative and the max might be an outlier.
4. The column total_income has some missing values.

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Good observation. </div>

In [110]:
# Get info on data
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [111]:
data.isna().sum()/data.shape[0]

children           0.000
days_employed      0.101
dob_years          0.000
education          0.000
education_id       0.000
family_status      0.000
family_status_id   0.000
gender             0.000
income_type        0.000
debt               0.000
total_income       0.101
purpose            0.000
dtype: float64

As we can see, there are missing values in days_employed and total_income columns.
The amount are the same, it will be intersting to check if it's on the same rows.
Also, we can see the the percentage in low, so it might not be so intersting.

In [112]:
for i in data:
    if data[i].isna().sum()>0:
        print(i)

days_employed
total_income


In [113]:
# Let's look in the filtered table at the the first column with missing data
data[data.days_employed.isna()]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

In [114]:
# Let's apply multiple conditions for filtering data and look at the number of rows in the filtered table.
data[(data.days_employed.isna())&(data.total_income.isna())]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

As we can see the total amount of missing values in both colums is 2174. The missing values seem to be symmetric.

**Intermediate conclusion**

The number of rows in the filtered table are the same as the number of missing values.
As conclusion we can assume that maybe these people aren't working (no days_employed and no total_income), but it could also be a mistake.
We can see that the percentage of the missing values are very small and the days_employed has messy values, we dont need this columns for our research so it's not necessary for now, I can make the values absolute.
As for the total_income, the percentage is low 1% of the data. We can fill the mean or median  income for each income type to be on the safe side.

Now, I would like to check if the missing data could be due to the specific client characteristic.
I chose to check the income type column - maybe there is a correlation between the type and the total income.

Let's check whether missing values are due to some column:

In [115]:
# Let's investigate clients who do not have data on identified characteristic and the column with the missing values
#data['income_type'].value_counts()
data[data.total_income.isna()]['income_type'].value_counts() 

employee         1105
business          508
retiree           413
civil servant     147
entrepreneur        1
Name: income_type, dtype: int64

In [116]:
data[data.total_income.isna()]['income_type'].value_counts()/len(data)

employee        0.051
business        0.024
retiree         0.019
civil servant   0.007
entrepreneur    0.000
Name: income_type, dtype: float64

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Great idea. You can also use `value_counts(normalize=True)`. Here is what I mean: </div>

In [117]:
# Reviewer's code

print('Nan dataframe. Income type distribution:')
display(data[data.total_income.isna()]['income_type'].value_counts())

print('Not Nan dataframe. Income type distribution:')
display(data[~data.total_income.isna()]['income_type'].value_counts())

Nan dataframe. Income type distribution:


employee         1105
business          508
retiree           413
civil servant     147
entrepreneur        1
Name: income_type, dtype: int64

Not Nan dataframe. Income type distribution:


employee                       10014
business                        4577
retiree                         3443
civil servant                   1312
unemployed                         2
student                            1
entrepreneur                       1
paternity / maternity leave        1
Name: income_type, dtype: int64

In [118]:
# Reviewer's code

print('Nan dataframe. Income type normalized distribution:')
display(data[data.total_income.isna()]['income_type'].value_counts(normalize=True))

print('Not Nan dataframe. Income type normalized distribution:')
display(data[~data.total_income.isna()]['income_type'].value_counts(normalize=True))

Nan dataframe. Income type normalized distribution:


employee        0.508
business        0.234
retiree         0.190
civil servant   0.068
entrepreneur    0.000
Name: income_type, dtype: float64

Not Nan dataframe. Income type normalized distribution:


employee                      0.517
business                      0.237
retiree                       0.178
civil servant                 0.068
unemployed                    0.000
student                       0.000
entrepreneur                  0.000
paternity / maternity leave   0.000
Name: income_type, dtype: float64

<div class="alert alert-info">
<b> Student's comment</b>

Nice I like the normalize, Thank you!
</div>



<font color='red'></font>    

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

Yes, it's a useful thing. 
</div>

In [119]:
# Checking the distribution in the whole dataset
for i in data:
    print(data[data.total_income.isna()][i].value_counts())
    print('-------------')


 0     1439
 1      475
 2      204
 3       36
 20       9
 4        7
-1        3
 5        1
Name: children, dtype: int64
-------------
Series([], Name: days_employed, dtype: int64)
-------------
34    69
40    66
31    65
42    65
35    64
36    63
47    59
41    59
30    58
28    57
57    56
58    56
54    55
38    54
56    54
37    53
52    53
39    51
33    51
50    51
51    50
45    50
49    50
29    50
43    50
46    48
55    48
48    46
53    44
44    44
60    39
61    38
62    38
64    37
32    37
27    36
23    36
26    35
59    34
63    29
25    23
24    21
66    20
65    20
21    18
22    17
67    16
0     10
68     9
69     5
20     5
71     5
70     3
72     2
19     1
73     1
Name: dob_years, dtype: int64
-------------
secondary education    1408
bachelor's degree       496
SECONDARY EDUCATION      67
Secondary Education      65
some college             55
Bachelor's Degree        25
BACHELOR'S DEGREE        23
primary education        19
Some College              7
S

After checking all columns, I can't see any correlation between the missing values and other category. The values are missing randomly.
I can't see any patterns in the missing data, I assume it's randomly.
Maybe people that didn't want to share thier income or there was a technical issue while they filled the form or either they are not working and have none income.

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Yes, they seem to be random. </div>

## Data transformation

Let's go through each column to see what issues we may have in them.

### Checking the data the `education` column:

In [120]:
# Let's see all values in education column to check if and what spellings will need to be fixed
print(data['education'].nunique())
print('-------')
data['education'].value_counts()

15
-------


secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

In [121]:
# Fix the registers if required
data['education'] = data['education'].str.lower()

In [122]:
# Checking all the values in the column to make sure we fixed them
print(data['education'].nunique())
print('-------')
data['education'].value_counts()

5
-------


secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

I can see that the education category has same name that written in lower or upper case.
Let's fix it and see how many categories we have.
After fixing the data,we have 5 unique education categories.

### Checking the data the `children` column:

In [123]:
# Let's see the distribution of values in the `children` column
data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [124]:
#pd.set_option('display.float_format', lambda x: f'{x:.3f}')
data.groupby('children')['family_status'].count()

children
-1        47
 0     14149
 1      4818
 2      2055
 3       330
 4        41
 5         9
 20       76
Name: family_status, dtype: int64

In [125]:
data['children'].value_counts()/len(data)

 0    0.657
 1    0.224
 2    0.095
 3    0.015
 20   0.004
-1    0.002
 4    0.002
 5    0.000
Name: children, dtype: float64

I can see that the data has negative amount of children (-1) and outlier of 20 children.
I asume the the client enter it by mistake. so I decide to change the -1 to 1, and 20 to 2.

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Agree. Considering the lack of clients with the number of children from 6 to 19, I suppose that 20 is a mistake. It could be just 2 or Nan that was trasformed into an anomaly when the table was downloaded or, which is more likely, merged with another table. 
</div>

In [126]:
# [fix the data based on your decision]
data['children'] = data.apply(lambda row: 2 if row['children']== 20  else row['children'],axis=1)
data['children'] = data.apply(lambda row: 1 if row['children']== -1 else row['children'],axis=1)
#can I do it in one line??

#another option:
#data.loc[data['children']== 20, ['children']] = 2
#data.loc[data['children']== -1, ['children']] = 1

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

A simple **replace** will do fine:

</div>

```python

vals = {-1: 1, 20:2}

data['children'] = data['children'].replace(vals)
```

In [127]:
# Checking the `children` column again to make sure it's all fixed
data['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

In [128]:
data['children'].value_counts()/len(data)

0   0.657
1   0.226
2   0.099
3   0.015
4   0.002
5   0.000
Name: children, dtype: float64

In [129]:
data.groupby('children')['family_status'].count()

children
0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: family_status, dtype: int64

We can see that it didn't affect the percentage so much.

### Checking the data the `days_employed` column:

In [130]:
# Find problematic data in `days_employed`, if they exist, and calculate the percentage
data['days_employed'].value_counts()/len(data)

-327.686    0.000
-1580.623   0.000
-4122.461   0.000
-2828.238   0.000
-2636.091   0.000
             ... 
-7120.518   0.000
-2146.884   0.000
-881.455    0.000
-794.666    0.000
-3382.114   0.000
Name: days_employed, Length: 19351, dtype: float64

It looks like the all data is corrupt , days cant have negative values. I thing the best way to handle this is to make the values absult. Any way this information is not nessecery for our main goal.

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Sounds reasonable, negative values seem to be normal apart from these minuses.
</div>

In [131]:
# Address the problematic values, if they exist
data['days_employed'] = data['days_employed'].abs()

In [132]:
# Check the result - make sure it's fixed
data['days_employed']

0         8437.673
1         4024.804
2         5623.423
3         4124.747
4       340266.072
           ...    
21520     4529.317
21521   343937.404
21522     2113.347
21523     3112.482
21524     1984.508
Name: days_employed, Length: 21525, dtype: float64

### Checking the data the `dob_years` column:

In [133]:
# Check the `dob_years` for suspicious values and count the percentage
data['dob_years'].value_counts()/data.shape[0]*100

35   2.866
40   2.829
41   2.820
34   2.801
38   2.778
42   2.774
33   2.699
39   2.662
31   2.602
36   2.578
44   2.541
29   2.532
30   2.509
48   2.499
37   2.495
50   2.388
43   2.383
32   2.369
49   2.360
28   2.337
45   2.309
27   2.290
56   2.262
52   2.249
47   2.230
54   2.225
46   2.207
58   2.142
57   2.137
53   2.132
51   2.081
59   2.063
55   2.058
26   1.895
60   1.751
25   1.659
61   1.649
62   1.635
63   1.250
64   1.231
24   1.226
23   1.180
65   0.901
66   0.850
22   0.850
67   0.776
21   0.516
0    0.469
68   0.460
69   0.395
70   0.302
71   0.269
20   0.237
72   0.153
19   0.065
73   0.037
74   0.028
75   0.005
Name: dob_years, dtype: float64

I can see the data has 0.469% of '0' in the age column. this is impossible.
I checked all data for '0' to see if I missed something.
I can see the dob_years is the only problematic category.
I chose to replace it with the median.

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Fortunately, there are only 101 values. If we had more, it would not be a good idea to replace them with the median. The median age is about 40, right? So we can receive 100 40 years old students or 100 40 years old employees with a few months of work experience. It may cause some questions. Another approach here is to take median according to job experience.
    
    
**A piece of advice:** before replacing values with median or whatever, make sure that they arose accidentally. Sometimes Nan-s and outliers carry important information. 

</div>

<div class="alert alert-info">
<b> Student's comment</b>

So it's better to earase them?
</div>



<font color='red'></font>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

It's better to leave them. I would say that if some values do not hurt the analysis, we can just ignore them. 
</div>

In [134]:
# Address the issues in the `dob_years` column, if they exist
for i in data:
    print(i, len(data[data[i]==0]))

children 14149
days_employed 0
dob_years 101
education 0
education_id 5260
family_status 0
family_status_id 12380
gender 0
income_type 0
debt 19784
total_income 0
purpose 0


There are '0' in 4 columns:
1. children - correct, some people don't have children.
2. dob_years - incorrect values, need to be fix.
3. education_id - it's ok is just the id no.
4. family_status_id - it's ok is just the id no.

In [135]:
data['dob_years'].median()

42.0

In [136]:
data['dob_years'].mean()

43.29337979094077

In [137]:
data[data['dob_years']==0]

children  days_employed  dob_years            education  education_id  \
99            0     346541.619          0  secondary education             1   
149           0       2664.273          0  secondary education             1   
270           3       1872.663          0  secondary education             1   
578           0     397856.565          0  secondary education             1   
1040          0       1158.030          0    bachelor's degree             0   
...         ...            ...        ...                  ...           ...   
19829         0            NaN          0  secondary education             1   
20462         0     338734.869          0  secondary education             1   
20577         0     331741.271          0  secondary education             1   
21179         2        108.967          0    bachelor's degree             0   
21313         0       1268.488          0  secondary education             1   

      family_status  family_status_id gender income_type  debt  total_income  \
99          married                 0      F     retiree     0     11406.644   
149        divorced                 3      F    employee     0     11228.230   
270         married                 0      F    employee     0     16346.633   
578         married                 0      F     retiree     0     15619.310   
1040       divorced                 3      F    business     0     48639.062   
...             ...               ...    ...         ...   ...           ...   
19829       married                 0      F    employee     0           NaN   
20462       married                 0      F     retiree     0     41471.027   
20577     unmarried                 4      F     retiree     0     20766.202   
21179       married                 0      M    business     0     38512.321   
21313     unmarried                 4      M    employee     0     24387.070   

                            purpose  
99                              car  
149            housing transactions  
270              housing renovation  
578    construction of own property  
1040                   to own a car  
...                             ...  
19829                       housing  
20462      purchase of my own house  
20577                      property  
21179        building a real estate  
21313          housing transactions  

[101 rows x 12 columns]

***replacing the '0' 'in dob_years' column:***

In [138]:
# Check the result - make sure it's fixed
data['dob_years'] = data['dob_years'].replace(0,data['dob_years'].median())

In [139]:
for i in data:
    print(i, len(data[data[i]==0]))

children 14149
days_employed 0
dob_years 0
education 0
education_id 5260
family_status 0
family_status_id 12380
gender 0
income_type 0
debt 19784
total_income 0
purpose 0


dob_years is fixed. all values are correct now.

### Checking the data the `family_status` column:

In [140]:
# Let's see the values for the column
data['family_status'].unique()

array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

All looks ok!

### Checking the data the `gender` column:

In [141]:
# Let's see the values in the column
data['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [142]:
print('XNA income is:',data[data['gender']== 'XNA']['total_income'].sum())
print('-----------')
print('Female max income is:',data[data['gender']== 'F']['total_income'].max())
print('-----------')
print('Male max income is:',data[data['gender']== 'M']['total_income'].max())
print('-----------')

XNA income is: 32624.825
-----------
Female max income is: 274402.943
-----------
Male max income is: 362496.645
-----------


I decide to change the 'XNA' to M because his income is higher the female max's income:

In [143]:
# Address the problematic values, if they exist
data.loc[data['gender']== 'XNA', 'gender']= 'F'

In [144]:
# Check the result - make sure it's fixed
data['gender'].unique()

array(['F', 'M'], dtype=object)

### Checking the data the `income_type` column:

In [145]:
# Let's see the values in the column
data['income_type'].unique()

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

No problem here.

# Checking for duplicates:

In [146]:
# Checking duplicates
data.duplicated().sum()

72

In [147]:
data[data.duplicated()].head()

children  days_employed  dob_years            education  education_id  \
2849         0            NaN     41.000  secondary education             1   
3290         0            NaN     58.000  secondary education             1   
4182         1            NaN     34.000    bachelor's degree             0   
4851         0            NaN     60.000  secondary education             1   
5557         0            NaN     58.000  secondary education             1   

          family_status  family_status_id gender income_type  debt  \
2849            married                 0      F    employee     0   
3290  civil partnership                 1      F     retiree     0   
4182  civil partnership                 1      F    employee     0   
4851  civil partnership                 1      F     retiree     0   
5557  civil partnership                 1      F     retiree     0   

      total_income                              purpose  
2849           NaN  purchase of the house for my family  
3290           NaN                    to have a wedding  
4182           NaN                     wedding ceremony  
4851           NaN                     wedding ceremony  
5557           NaN                    to have a wedding

In [148]:
# Address the duplicates, if they exist
data_1 = data.copy()
data_1 = data.drop_duplicates().reset_index(drop=True)

In [149]:
# Last check whether we have any duplicates
data_1.duplicated().sum()

0

In [150]:
# Check the size of the dataset that you now have after your first manipulations with it
print(data.shape)
print('-------')
print(data_1.shape)
print('-------')
(1-data_1.shape[0]/data.shape[0])*100

(21525, 12)
-------
(21453, 12)
-------


0.33449477351916723

We droped 72 duplicated rows and we can see now that the data is clean , the duplicates were only 0.33% of the data.
for my convenience I will change the data_1 to data.

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
It is better to use **copy** method when copying dataframe. 

</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

As there are no any unique id numbers for the clients, these duplicates can be related to different people. 

</div>

<div class="alert alert-info">
<b> Student's comment</b>

Ok noted, thanks.
I used it :) hope this is what you ment.
</div>
<font color='red'></font>     

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

It is not a mistake for now, but in the future it cause some problems. Let me show you an example:
</div>

In [151]:
# Reviewer's code 2.1

l = {'A':[1,2,4], 'B':[3,3,5]}

df1 = pd.DataFrame(l)

df2 = df1

df3 = df1.copy() 

display(df1, df2, df3)

A  B
0  1  3
1  2  3
2  4  5

A  B
0  1  3
1  2  3
2  4  5

A  B
0  1  3
1  2  3
2  4  5

In [152]:
# Reviewer's code 2.2

df2.loc[0, 'A'] = 'change2'

df3.loc[1, 'A'] = 'change3'

display(df1, df2, df3)

A  B
0  change2  3
1        2  3
2        4  5

A  B
0  change2  3
1        2  3
2        4  5

A  B
0        1  3
1  change3  3
2        4  5

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

As we see, the old dataframe, **df1** has the same changes as **df2** does, because **df2** was created without using **copy**. While the **df3** changes did not accur in the **df1** frame as **df3** was created with **copy**. 

</div>

In [153]:
data = data_1

In [154]:
data.shape

(21453, 12)

# Working with missing values

It's better to work with dictionaries because it will be easier to address the values and it will save time.

In [155]:
# Find the dictionaries
data[['education','education_id','family_status', 'family_status_id']]

education  education_id      family_status  family_status_id
0        bachelor's degree             0            married                 0
1      secondary education             1            married                 0
2      secondary education             1            married                 0
3      secondary education             1            married                 0
4      secondary education             1  civil partnership                 1
...                    ...           ...                ...               ...
21448  secondary education             1  civil partnership                 1
21449  secondary education             1            married                 0
21450  secondary education             1  civil partnership                 1
21451  secondary education             1            married                 0
21452  secondary education             1            married                 0

[21453 rows x 4 columns]

### Restoring missing values in `total_income`

We have 2 columns with missing values:
1. days_employed
2. total_income

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Do we need a loop here?
</div>

<div class="alert alert-info">
<b> Student's comment</b>

I guess not because I already know the columns.
So I removed it.
</div>
<font color='red'></font>     


In [156]:
data['dob_years'].value_counts().sort_index()

19.000     14
20.000     51
21.000    111
22.000    183
23.000    252
24.000    264
25.000    357
26.000    408
27.000    493
28.000    503
29.000    544
30.000    537
31.000    559
32.000    509
33.000    581
34.000    601
35.000    616
36.000    554
37.000    536
38.000    597
39.000    572
40.000    607
41.000    605
42.000    696
43.000    512
44.000    545
45.000    496
46.000    472
47.000    477
48.000    536
49.000    508
50.000    513
51.000    446
52.000    484
53.000    459
54.000    476
55.000    443
56.000    483
57.000    456
58.000    454
59.000    443
60.000    374
61.000    354
62.000    348
63.000    269
64.000    260
65.000    193
66.000    182
67.000    167
68.000     99
69.000     85
70.000     65
71.000     56
72.000     33
73.000      8
74.000      6
75.000      1
Name: dob_years, dtype: int64

In [157]:
# Let's write a function that calculates the age category
def age_category(age):
    if age < 19:
        return 'too young for loan'
    elif age < 29:
        return '20-29 years old'
    elif age < 39:
        return '30-39 years old'
    elif age < 49:
        return '40-49 years old'
    elif age < 59:
        return '50-59 years old'
    elif age < 69:
        return '60-69 years old'
    else:
        return '70+ years old'

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Nice idea. 
</div>

In [158]:
# Test if the function works
age_category(55)

'50-59 years old'

In [159]:
# Creating new column based on function
data['age_group'] = data['dob_years'].apply(age_category)

In [160]:
# Checking how values in the new column
data[['dob_years','age_group']].head()

dob_years        age_group
0     42.000  40-49 years old
1     36.000  30-39 years old
2     33.000  30-39 years old
3     32.000  30-39 years old
4     53.000  50-59 years old

***Let's check the missing values now:***

In [161]:
data.isna().sum()

children               0
days_employed       2102
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2102
purpose                0
age_group              0
dtype: int64

In [162]:
data[(data.total_income.isna()) & (data.days_employed.isna())]

children  days_employed  dob_years            education  education_id  \
12            0            NaN     65.000  secondary education             1   
26            0            NaN     41.000  secondary education             1   
29            0            NaN     63.000  secondary education             1   
41            0            NaN     50.000  secondary education             1   
55            0            NaN     54.000  secondary education             1   
...         ...            ...        ...                  ...           ...   
21417         2            NaN     47.000  secondary education             1   
21423         1            NaN     50.000  secondary education             1   
21425         0            NaN     48.000    bachelor's degree             0   
21430         1            NaN     42.000  secondary education             1   
21438         2            NaN     28.000  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21417            married                 0      M       business     0   
21423  civil partnership                 1      F       employee     0   
21425            married                 0      F       business     0   
21430            married                 0      F       employee     0   
21438            married                 0      F       employee     0   

       total_income                   purpose        age_group  
12              NaN         to have a wedding  60-69 years old  
26              NaN                 education  40-49 years old  
29              NaN    building a real estate  60-69 years old  
41              NaN  second-hand car purchase  50-59 years old  
55              NaN         to have a wedding  50-59 years old  
...             ...                       ...              ...  
21417           NaN         purchase of a car  40-49 years old  
21423           NaN          wedding ceremony  50-59 years old  
21425           NaN       building a property  40-49 years old  
21430           NaN    building a real estate  40-49 years old  
21438           NaN              car purchase  20-29 years old  

[2102 rows x 13 columns]

As we can see the total amount of missing values in both colums is 2102.
we can assume that all the missing values appears on the same rows.

In [163]:
# Create a table without missing values and print a few of its rows to make sure it looks fine
data[data.total_income.notna()].head(10)

children  days_employed  dob_years            education  education_id  \
0         1       8437.673     42.000    bachelor's degree             0   
1         1       4024.804     36.000  secondary education             1   
2         0       5623.423     33.000  secondary education             1   
3         3       4124.747     32.000  secondary education             1   
4         0     340266.072     53.000  secondary education             1   
5         0        926.186     27.000    bachelor's degree             0   
6         0       2879.202     43.000    bachelor's degree             0   
7         0        152.780     50.000  secondary education             1   
8         2       6929.865     35.000    bachelor's degree             0   
9         0       2188.756     41.000  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose        age_group  
0                purchase of the house  40-49 years old  
1                         car purchase  30-39 years old  
2                purchase of the house  30-39 years old  
3              supplementary education  30-39 years old  
4                    to have a wedding  50-59 years old  
5                purchase of the house  20-29 years old  
6                 housing transactions  40-49 years old  
7                            education  50-59 years old  
8                     having a wedding  30-39 years old  
9  purchase of the house for my family  40-49 years old

In [164]:
# Look at the mean values for income based on your identified factors
for i in data[['income_type','education','gender','family_status']]:
    print(data.groupby([i])['total_income'].mean())
    print('-------------')

income_type
business                      32386.794
civil servant                 27343.730
employee                      25820.842
entrepreneur                  79866.103
paternity / maternity leave    8612.661
retiree                       21940.395
student                       15712.260
unemployed                    21014.361
Name: total_income, dtype: float64
-------------
education
bachelor's degree     33142.802
graduate degree       27960.025
primary education     21144.882
secondary education   24594.503
some college          29045.444
Name: total_income, dtype: float64
-------------
gender
F   24656.230
M   30907.144
Name: total_income, dtype: float64
-------------
family_status
civil partnership   26694.429
divorced            27189.355
married             27041.785
unmarried           26934.070
widow / widower     22984.209
Name: total_income, dtype: float64
-------------


In [165]:
# Look at the median values for income based on your identified factors
for i in data[['income_type','education','gender','family_status']]:
    print(data.groupby([i])['total_income'].median())
    print('-------------')

income_type
business                      27577.272
civil servant                 24071.669
employee                      22815.103
entrepreneur                  79866.103
paternity / maternity leave    8612.661
retiree                       18962.318
student                       15712.260
unemployed                    21014.361
Name: total_income, dtype: float64
-------------
education
bachelor's degree     28054.531
graduate degree       25161.584
primary education     18741.976
secondary education   21836.583
some college          25618.464
Name: total_income, dtype: float64
-------------
gender
F   21465.165
M   26834.295
Name: total_income, dtype: float64
-------------
family_status
civil partnership   23186.534
divorced            23515.096
married             23389.540
unmarried           23149.028
widow / widower     20514.190
Name: total_income, dtype: float64
-------------


I decide to checke factors that can effect person's income.

income_type - variety of incomes.

education - variety of incomes.

gender - around 5-6K diffrence.

family_status - no a big diffrence.

I decide to fill the income missing values according to the median income type.
There are many income type and I assume it will be more accurate to use this category.
The median is lower than the mean, it means there are some outlier and this is why I choose the medain.

In [166]:
#  Write a function that we will use for filling in missing values
data['total_income'] = data['total_income'].fillna(data.groupby('income_type')['total_income'].transform('median'))

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

1. Very good decision. We can also add **education** in a groupby method.
    
    
2. What kind of function do you want to receive? I mean, it is just one line of code. do we really need a function? If you are talking about the task, where it's said about some function, then don't worry, since your code is elegant. No *def* function is needed anymore. Actually, **fillna** is a function, so everything is ok here. 
    
</div>

<div class="alert alert-info">
<b> Student's comment</b>
    
1. you mean double groupby? or instead of the income type?
    
2. Great thank you! I thought they ask for a def.
    
</div>
<font color='red'></font>     

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

I mean, you can use several columns to take a median value: 
    
    `data['total_income'].fillna(data.groupby(['income_type', 'education'])['total_income'].transform('median'))`


Then we will calculate a median **total_income** value for each pair **income_type-education**.</div>

In [167]:
# Check if it works
data.total_income.isna().sum()

0

In [168]:
# Check if we got any errors
data['total_income']

0       40620.102
1       17932.802
2       23341.752
3       42820.568
4       25378.572
           ...   
21448   35966.698
21449   24959.969
21450   14347.610
21451   39054.888
21452   13127.587
Name: total_income, Length: 21453, dtype: float64

In [169]:
# Checking the number of entries in the columns
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21453 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21453 non-null  float64
 3   education         21453 non-null  object 
 4   education_id      21453 non-null  int64  
 5   family_status     21453 non-null  object 
 6   family_status_id  21453 non-null  int64  
 7   gender            21453 non-null  object 
 8   income_type       21453 non-null  object 
 9   debt              21453 non-null  int64  
 10  total_income      21453 non-null  float64
 11  purpose           21453 non-null  object 
 12  age_group         21453 non-null  object 
dtypes: float64(3), int64(4), object(6)
memory usage: 2.1+ MB


###  Restoring values in `days_employed`

In [170]:
# Distribution of `days_employed` medians based on your identified parameters
days_employes_median = data.groupby('dob_years')['days_employed'].median()
days_employes_median

dob_years
19.000      724.493
20.000      674.839
21.000      618.734
22.000      703.310
23.000      690.204
24.000      947.731
25.000      919.199
26.000     1083.658
27.000     1166.212
28.000     1141.705
29.000     1315.454
30.000     1420.587
31.000     1308.901
32.000     1446.623
33.000     1426.415
34.000     1615.910
35.000     1613.495
36.000     1799.520
37.000     1816.713
38.000     1817.194
39.000     1891.389
40.000     1728.937
41.000     1864.658
42.000     2210.576
43.000     1869.020
44.000     2084.330
45.000     2254.431
46.000     2100.473
47.000     2203.079
48.000     2429.675
49.000     2560.317
50.000     2626.986
51.000     2846.081
52.000     3395.640
53.000     3650.008
54.000     4026.541
55.000     7043.471
56.000     6850.179
57.000   333953.492
58.000   331245.977
59.000   344372.337
60.000   349414.537
61.000   343984.932
62.000   354981.718
63.000   358482.296
64.000   355781.563
65.000   357103.967
66.000   363827.439
67.000   357222.855
68.000   3

In [171]:
# Distribution of `days_employed` means based on your identified parameters
days_employes_mean = data.groupby('dob_years')['days_employed'].mean()
days_employes_mean

dob_years
19.000      633.678
20.000      684.944
21.000      709.441
22.000     2793.322
23.000      827.309
24.000     1026.405
25.000     1088.406
26.000     3214.617
27.000     3725.821
28.000     2180.056
29.000     1553.823
30.000     1696.039
31.000     2331.247
32.000     3877.197
33.000     3241.413
34.000     4141.243
35.000     2755.226
36.000     4496.535
37.000     5879.263
38.000     7641.485
39.000     5277.456
40.000     7076.522
41.000     6400.926
42.000    17741.626
43.000     9418.428
44.000     9952.060
45.000    11748.039
46.000    14121.192
47.000    14223.686
48.000    18149.684
49.000    26679.791
50.000    50493.031
51.000    68425.773
52.000    82803.402
53.000    94073.249
54.000   126757.782
55.000   152264.947
56.000   157417.505
57.000   193689.068
58.000   188571.490
59.000   227496.974
60.000   262925.278
61.000   246731.049
62.000   274381.284
63.000   293383.721
64.000   287670.476
65.000   287448.539
66.000   313613.718
67.000   317629.789
68.000   3

In [172]:
diff_days_employes = days_employes_mean - days_employes_median
print(diff_days_employes.count())
diff_days_employes.sort_values()

57


dob_years
58.000   -142674.487
57.000   -140264.424
59.000   -116875.363
74.000   -101813.322
61.000    -97253.883
60.000    -86489.260
62.000    -80600.434
65.000    -69655.429
64.000    -68111.087
63.000    -65098.575
73.000    -50442.882
66.000    -50213.721
70.000    -44715.116
67.000    -39593.067
68.000    -38381.446
72.000    -34201.151
71.000    -26948.957
69.000    -23951.863
19.000       -90.815
75.000         0.000
20.000        10.105
24.000        78.674
21.000        90.707
23.000       137.105
25.000       169.207
29.000       238.370
30.000       275.452
31.000      1022.346
28.000      1038.351
35.000      1141.732
33.000      1814.998
22.000      2090.012
26.000      2130.959
32.000      2430.575
34.000      2525.333
27.000      2559.609
36.000      2697.014
39.000      3386.068
37.000      4062.550
41.000      4536.269
40.000      5347.585
38.000      5824.291
43.000      7549.409
44.000      7867.730
45.000      9493.608
47.000     12020.607
46.000     12020.718
42.

I decided to check the days according to age as it might effect the amount of days people work.
I choose the median because there is a big diffrent between median and mean, so we probably have some outliers.

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

How about age? I suppose, there's a stronger connection between days employed and age than between days employed and family status. What do you think? </div>

<div class="alert alert-info">
<b> Student's comment</b>
    
I thought that probably people with children migth work less :) 
But maybe other family status is not so relevant. so, I guess age is better! I changed it.
    
</div>
<font color='red'></font>     

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

Good :) 
</div>

In [173]:
# Let's write a function that calculates means or medians (depending on your decision) based on your identified parameter
data['days_employed'] = data['days_employed'].fillna(data.groupby('dob_years')['days_employed'].transform('median'))

In [174]:
# Check that the function works
data['days_employed'].isna().sum()

0

Let's check if the data is fine now:

In [175]:
# Check the entries in all columns - make sure we fixed all missing values
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21453 non-null  int64  
 1   days_employed     21453 non-null  float64
 2   dob_years         21453 non-null  float64
 3   education         21453 non-null  object 
 4   education_id      21453 non-null  int64  
 5   family_status     21453 non-null  object 
 6   family_status_id  21453 non-null  int64  
 7   gender            21453 non-null  object 
 8   income_type       21453 non-null  object 
 9   debt              21453 non-null  int64  
 10  total_income      21453 non-null  float64
 11  purpose           21453 non-null  object 
 12  age_group         21453 non-null  object 
dtypes: float64(3), int64(4), object(6)
memory usage: 2.1+ MB


All categories has 21453 rows and no missing values.

## Categorization of data

Let's categorize the peoples according to thier purpose:

In [176]:
# Print the values for your selected data for categorization
data[['children', 'family_status','purpose', 'total_income']]

children      family_status                  purpose  total_income
0             1            married    purchase of the house     40620.102
1             1            married             car purchase     17932.802
2             0            married    purchase of the house     23341.752
3             3            married  supplementary education     42820.568
4             0  civil partnership        to have a wedding     25378.572
...         ...                ...                      ...           ...
21448         1  civil partnership     housing transactions     35966.698
21449         0            married        purchase of a car     24959.969
21450         1  civil partnership                 property     14347.610
21451         3            married        buying my own car     39054.888
21452         2            married             to buy a car     13127.587

[21453 rows x 4 columns]

Let's check unique values:

In [177]:
# Check the unique values
print('there are',data.purpose.nunique(), 'unique purposes.')
print('--------')
data.purpose.unique()

there are 38 unique purposes.
--------


array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

There are some word that repeated it the unique purpose.
I will categorize them by:
1. house purpose (house, housing, real estate, property)
2. car purpose (car)
3. wedding purpose (wedding)
4. education purpose (education, educated, university)

[Based on these themes, we will probably want to categorize our data.]


In [178]:
from pymystem3 import Mystem
from collections import Counter
m = Mystem()

In [179]:
import nltk
from nltk.stem import WordNetLemmatizer

wordnet_lemma = WordNetLemmatizer()

In [180]:
lemmas_list_all = []

for issue in data.purpose.unique():
    words = nltk.word_tokenize(issue)
    lemmas = [wordnet_lemma.lemmatize(w, pos = 'n') for w in words]
    lemmas=[l.lower() for l in lemmas]
    for i in lemmas:
        lemmas_list_all.append(i)

In [181]:
lemmas_list_all

['purchase',
 'of',
 'the',
 'house',
 'car',
 'purchase',
 'supplementary',
 'education',
 'to',
 'have',
 'a',
 'wedding',
 'housing',
 'transaction',
 'education',
 'having',
 'a',
 'wedding',
 'purchase',
 'of',
 'the',
 'house',
 'for',
 'my',
 'family',
 'buy',
 'real',
 'estate',
 'buy',
 'commercial',
 'real',
 'estate',
 'buy',
 'residential',
 'real',
 'estate',
 'construction',
 'of',
 'own',
 'property',
 'property',
 'building',
 'a',
 'property',
 'buying',
 'a',
 'second-hand',
 'car',
 'buying',
 'my',
 'own',
 'car',
 'transaction',
 'with',
 'commercial',
 'real',
 'estate',
 'building',
 'a',
 'real',
 'estate',
 'housing',
 'transaction',
 'with',
 'my',
 'real',
 'estate',
 'car',
 'to',
 'become',
 'educated',
 'second-hand',
 'car',
 'purchase',
 'getting',
 'an',
 'education',
 'car',
 'wedding',
 'ceremony',
 'to',
 'get',
 'a',
 'supplementary',
 'education',
 'purchase',
 'of',
 'my',
 'own',
 'house',
 'real',
 'estate',
 'transaction',
 'getting',
 'higher'

In [182]:
Counter(lemmas_list_all)

Counter({'purchase': 6,
         'of': 5,
         'the': 2,
         'house': 3,
         'car': 9,
         'supplementary': 2,
         'education': 7,
         'to': 6,
         'have': 1,
         'a': 9,
         'wedding': 3,
         'housing': 3,
         'transaction': 4,
         'having': 1,
         'for': 2,
         'my': 4,
         'family': 1,
         'buy': 4,
         'real': 7,
         'estate': 7,
         'commercial': 2,
         'residential': 1,
         'construction': 1,
         'own': 4,
         'property': 4,
         'building': 2,
         'buying': 3,
         'second-hand': 2,
         'with': 2,
         'become': 1,
         'educated': 1,
         'getting': 2,
         'an': 1,
         'ceremony': 1,
         'get': 1,
         'higher': 1,
         'profile': 1,
         'university': 2,
         'renting': 1,
         'out': 1,
         'renovation': 1,
         'going': 1})

In [183]:
house_category = ['house', 'estate', 'housing','property']
car_category = ['car']
wedding_category = ['wedding']
education_category = ['education', 'educated', 'university']

In [184]:
def lemmatization_func(line):
  
    words = nltk.word_tokenize(line)
    lemmas = [wordnet_lemma.lemmatize(w, pos = 'n') for w in words]
    lemmas = [l.lower() for l in lemmas]
    
    if any(word in lemmas for word in house_category):
        return 'house purpose'
    elif  any(word in lemmas for word in car_category):
        return 'car purpose'
    elif  any(word in lemmas for word in wedding_category):
        return 'wedding purpose'
    elif  any(word in lemmas for word in education_category):
        return 'education purpose'
    else:
        return 'other'

In [185]:
# Create a column with the categories and count the values for them
data['clean_purpose'] = data['purpose'].apply(lemmatization_func)
data.head()

children  days_employed  dob_years            education  education_id  \
0         1       8437.673     42.000    bachelor's degree             0   
1         1       4024.804     36.000  secondary education             1   
2         0       5623.423     33.000  secondary education             1   
3         3       4124.747     32.000  secondary education             1   
4         0     340266.072     53.000  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose        age_group      clean_purpose  
0    purchase of the house  40-49 years old      house purpose  
1             car purchase  30-39 years old        car purpose  
2    purchase of the house  30-39 years old      house purpose  
3  supplementary education  30-39 years old  education purpose  
4        to have a wedding  50-59 years old    wedding purpose

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Great code and I agree with the categories. But we can make it simple: just check whether a particular string contains the word or a part of the word. Example: 
    
    
 
</div>

In [186]:
# Reviewer's code

print('hous' in 'housing')

print('hous' in 'house')

print('car' in 'to buy a car')

print('car' in 'housing')

True
True
True
False


<div class="alert alert-info">
<b> Student's comment</b>
    
I am not sure what the code you wrote can replace in my code :)
    
</div>
<font color='red'></font>  


<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

We can write something like this: 
</div>

In [187]:
# Reviewer's code

def func_purpose(row):
    try:
        if 'wed' in row:
            return 'wedding'
        elif 'estat' in row or 'hous' in row:
            return 'real_estate'
        elif 'car' in row:
            return 'car'
        else:
            return 'education'
    except:
        return 'no_category'
    
    
# Check the unique values after we apply the function to a column:    
data['purpose'].apply(func_purpose).unique() 


array(['real_estate', 'car', 'education', 'wedding'], dtype=object)

In [188]:
data['clean_purpose'].value_counts()

house purpose        10811
car purpose           4306
education purpose     4013
wedding purpose       2323
Name: clean_purpose, dtype: int64

In [189]:
data['clean_purpose'].value_counts()/data.shape[0]*100

house purpose       50.394
car purpose         20.072
education purpose   18.706
wedding purpose     10.828
Name: clean_purpose, dtype: float64

In [190]:
pd.set_option('display.float_format', lambda x: f'{x:.3f}')
data.groupby('clean_purpose')['total_income'].sum()

clean_purpose
car purpose         113831851.666
education purpose   104416685.037
house purpose       288351206.258
wedding purpose      60861764.112
Name: total_income, dtype: float64

50% of the people are planning to buy a house with the loan.

The people that wants to buy a house also have the higest income.

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

We are asked to pay attention to dictionaries. A pair of columns `education` and `education_id` can be called a dictionary, since they are interchangeable and a dictionary (dict data structure) can be made with them. Is there anything else like this in the dataset? 

</div>

<div class="alert alert-info">
<b> Student's comment</b>

I printed all the dic columns above.
- education_id
- education
- family_status_id
- family_status
    
I didn't know where to use the dictionary...
In whitch task?

</div>
<font color='red'></font>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
True. We can create them and save in another dataframe: 
</div>

In [191]:
# Reviewer's code 2

data[['family_status_id', 'family_status']].drop_duplicates().reset_index(drop=True)

family_status_id      family_status
0                 0            married
1                 1  civil partnership
2                 2    widow / widower
3                 3           divorced
4                 4          unmarried

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
Unfortunately, the purpose of creating them is not stated in the task, so I'll show you the example in the hypotheses part.</div>

## Checking the Hypotheses


**Is there a correlation between having children and paying back on time?**

In [192]:
# Check the children data and paying back on time

# Calculating default-rate based on the number of children

def my_mean(x):
    return '{:.2%} '.format(x.mean())

data.groupby('children')['debt'].agg(['count', 'sum', my_mean])

count   sum my_mean
children                     
0         14090  1063  7.54% 
1          4855   445  9.17% 
2          2128   202  9.49% 
3           330    27  8.18% 
4            41     4  9.76% 
5             9     0  0.00%

**Conclusion**

People with no children tends to ask more for a loan than others, 7.5% on them don't pay of time.
The less children you have the higher probability you will ask for a loan.
People with 1-4 children also have some diffiuclties with the payments, 8-10% of them don't pay on time.

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

But sample with 5 kids is too small to make any conclusions :) 
</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

You're right. In the column `debt` 1 indicates the presence of debt. Zero is its absence. So, the sum of the whole column will be equal to the sum of 1-s or the number of debtors, and applying the **count** method will return the total number of loans. Thus, to find the share of debtors, you need to divide this sum by the number of loans calculated with count method, and this is nothing more than just mean.</div>

<div class="alert alert-info">
<b> Student's comment</b>
    
I deleted the conclusion about the 5 kids.

</div>
<font color='red'></font>

In [193]:
# Reviewer's code

data.groupby('children')['debt'].agg(['count', 'sum', 'mean'])

count   sum  mean
children                   
0         14090  1063 0.075
1          4855   445 0.092
2          2128   202 0.095
3           330    27 0.082
4            41     4 0.098
5             9     0 0.000

In [194]:
# Reviewer's code

def my_mean(x): return x.mean()*100

# We can you our own function in agg method
data.groupby('children')['debt'].agg(['count', 'sum', my_mean])

count   sum  my_mean
children                      
0         14090  1063    7.544
1          4855   445    9.166
2          2128   202    9.492
3           330    27    8.182
4            41     4    9.756
5             9     0    0.000

In [195]:
# Reviewer's code

def my_mean(x): return '{:.2%} '.format(x.mean())

data.groupby('children')['debt'].agg(['count', 'sum', my_mean])

count   sum my_mean
children                     
0         14090  1063  7.54% 
1          4855   445  9.17% 
2          2128   202  9.49% 
3           330    27  8.18% 
4            41     4  9.76% 
5             9     0  0.00%

**Is there a correlation between family status and paying back on time?**

In [196]:
def my_mean(x):
    return '{:.2%} '.format(x.mean())

data.groupby('family_status')['debt'].agg(['count', 'sum', my_mean])


count  sum my_mean
family_status                        
civil partnership   4150  388  9.35% 
divorced            1195   85  7.11% 
married            12339  931  7.55% 
unmarried           2810  274  9.75% 
widow / widower      959   63  6.57%

**Conclusion**

The most irresponsible are those who have never been married.

All family status has 6.5-9.7% probability to not paying on time.
unmarried and civil partnership tends to have more debts than others.
widow/widower has less debts than others.
Married people ask more for a loan than others, 7.5% on them don't pay on time.

<div class="alert alert-info">
<b> Student's comment</b>
    
Thank you!
I like this code it's super elegant.

</div>
<font color='red'></font>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Widowed people are, as a rule, older, and they are more responsible. And most likely they already have adult children who can help with payments. In general, we can say this: the most irresponsible are those who have never been married. </div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

All right, we have pivot tables. The question about the dictionaries is based on the idea that 1. we don't need interchangeable columns in the dataset and 2. we should use numbers instead of str values when using some machine learning algorithms. Let's assume that we drop the **family_status** column after creating the dictionary, leaving only **family_status_id** in the dataset. Then we want to display the debt statistics. In order to make it clear we use dictionary to replace number with str value. That's all. So it's just a good example of using dicts for replacing values in the dataframe. And it is also a good example of saving these values in a small data structure.</div>

In [197]:
# Reviewer's code 2

dict(zip(data['family_status_id'],data['family_status']))

{0: 'married',
 1: 'civil partnership',
 2: 'widow / widower',
 3: 'divorced',
 4: 'unmarried'}

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
The dictionary will look like this. But I'll create it as a dataframe so that it is clear that the same can be done with dataframes just by taking a slice of two columns.
</div>

In [199]:
# Reviewer's code

# Create a dictionary
family_dict = data[['family_status_id', 'family_status']]
family_dict = family_dict.drop_duplicates().reset_index(drop=True)
family_dict

def my_mean(x): return '{:.2%} '.format(x.mean())

# Table as in the example above
a = data.groupby('family_status_id')['debt'].agg(['count', 'sum', my_mean])
display(a)

# Replacement
a.reset_index().replace({'family_status_id': family_dict.family_status.to_dict()})

count  sum my_mean
family_status_id                    
0                 12339  931  7.55% 
1                  4150  388  9.35% 
2                   959   63  6.57% 
3                  1195   85  7.11% 
4                  2810  274  9.75%

family_status_id  count  sum my_mean
0            married  12339  931  7.55% 
1  civil partnership   4150  388  9.35% 
2    widow / widower    959   63  6.57% 
3           divorced   1195   85  7.11% 
4          unmarried   2810  274  9.75%

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

This is a great example of how you can replace data using a dictionary key :)
</div>

**Is there a correlation between income level and paying back on time?**

In [97]:
# categorize income in function:
def income_category(income):
    if income < 10000:
        return '0-10K'
    elif income < 50000:
        return '10K-50K'
    elif income < 100000:
        return '50K-100K'
    elif income < 150000:
        return '100K-150K'
    elif income < 200000:
        return '150K-200K'
    elif income < 250000:
        return '200K-250K'
    elif income < 300000:
        return '250K-300K'
    else:
        return '300K +'

In [98]:
#apply the function on the data:
data['income_cat'] = data['total_income'].apply(income_category)

In [99]:
# Check the income level data and paying back on time

# Calculating default-rate based on income level

def my_mean(x):
    return '{:.2%} '.format(x.mean())

data.groupby('income_cat')['debt'].agg(['count', 'sum', my_mean])



count   sum  my_mean
income_cat                      
0-10K         926    58   6.26% 
100K-150K      71     4   5.63% 
10K-50K     19206  1591   8.28% 
150K-200K      17     1   5.88% 
200K-250K       5     0   0.00% 
250K-300K       4     0   0.00% 
300K +          2     1  50.00% 
50K-100K     1222    86   7.04%

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
The destribution is not fair. Please, check out their sizes :) </div>

<div class="alert alert-info">
<b> Student's comment</b>
    
now it's ok?
</div>
<font color='red'></font>

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

10K-50K is 19206 size, while others are very small. It's not fair to compare them. 

One possible solution is to divide the column into 5-7 approx. equal categories. For example, if we have 10000 rows and we want to get 5 groups, we can divide them with the step 2000. It is not the best idea though, but it is stille better than analysing a skewed distribution.  It can be implemented with a simple loop or list generator. A **numpy** has also the [array_split](https://numpy.org/doc/stable/reference/generated/numpy.array_split.html) method. 
    
There's another way, a pandas **qcut** method that can be quite useful. It can be tricky at the beginning, so I'll show you an example below. The idea behind qcut is that it splits the list of values according to quintiles. </div>

In [100]:
# Reviewer's code 2

pd.qcut(data['total_income'], q=5)

0        (34339.686, 362496.645]
1         (15765.943, 21141.071]
2         (21141.071, 25814.809]
3        (34339.686, 362496.645]
4         (21141.071, 25814.809]
                  ...           
21448    (34339.686, 362496.645]
21449     (21141.071, 25814.809]
21450      (3306.761, 15765.943]
21451    (34339.686, 362496.645]
21452      (3306.761, 15765.943]
Name: total_income, Length: 21453, dtype: category
Categories (5, interval[float64]): [(3306.761, 15765.943] < (15765.943, 21141.071] < (21141.071, 25814.809] < (25814.809, 34339.686] < (34339.686, 362496.645]]

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment</h2>
    

Fast and simple :) You can take as many groups as you want. Let's try 3: </div>

In [101]:
# Reviewer's code 2

pd.qcut(data['total_income'], q=3)

0        (27577.272, 362496.645]
1          (3306.761, 19081.788]
2         (19081.788, 27577.272]
3        (27577.272, 362496.645]
4         (19081.788, 27577.272]
                  ...           
21448    (27577.272, 362496.645]
21449     (19081.788, 27577.272]
21450      (3306.761, 19081.788]
21451    (27577.272, 362496.645]
21452      (3306.761, 19081.788]
Name: total_income, Length: 21453, dtype: category
Categories (3, interval[float64]): [(3306.761, 19081.788] < (19081.788, 27577.272] < (27577.272, 362496.645]]

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

Now it returned 3 categories. 


Then we can give them names: </div>

In [102]:
# Reviewer's code 2

pd.qcut(data['total_income'], 5, ['low', 'below average', 'average', 'above average', 'high'])

0                 high
1        below average
2              average
3                 high
4              average
             ...      
21448             high
21449          average
21450              low
21451             high
21452              low
Name: total_income, Length: 21453, dtype: category
Categories (5, object): ['low' < 'below average' < 'average' < 'above average' < 'high']

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

Much better. Another good thing about this method is that we can send it into groupby method: </div>

In [103]:
# Reviewer's code 2

cats = pd.qcut(data['total_income'], 5, ['low', 'below average', 'average', 'above average', 'high'])

data.groupby(cats)['debt'].agg(['count', 'sum', lambda x: '{:.2%} '.format(x.mean())])

count  sum <lambda_0>
total_income                        
low             4291  344     8.02% 
below average   4290  361     8.41% 
average         4291  375     8.74% 
above average   4290  361     8.41% 
high            4291  300     6.99%

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

So, we group by the categories created with qcut. 


Again, it is not the best solution - to divide into equal parts. The best idea here would be gather more information about the dataframe :) In real life it is ok to ask some questions about data that is given to us.  </div>

**Conclusion**

People with total income of 10K-50K tends to ask more for a loan than others, 8.2% of them didn't pay on time.
After them are people with total income of 50K-100K , 7% of them didn't pay on time.

Only two people earn more than 300K and one of them didn't return the loan on time.
9 people with total income of 200K-300K paid one time!

The rest (0-10K , 100K-200K) has 5-6% default rate.

**How does credit purpose affect the default rate?**

In [104]:
def my_mean(x):
    return '{:.2%} '.format(x.mean())

data.groupby('clean_purpose')['debt'].agg(['count', 'sum', my_mean])



count  sum my_mean
clean_purpose                        
car purpose         4306  403  9.36% 
education purpose   4013  370  9.22% 
house purpose      10811  782  7.23% 
wedding purpose     2323  186  8.01%

**Conclusion**

most of the people are asking for a loan in order to buy a house and relatively they tend to pay back on time more than the others.

9% of people who ask a loan to buy a car or pay for education tends not to pay on time.

Asking for a loan to a wedding is less common than others and 8% of them didn't pay on time. 

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment </b>
    
Seems to be true.</div>

# General Conclusion 

**We had 2 columns with missing values:**
1. days_employed
2. total_income
the percentage of the missing values are very small 1%.

**We couldn’t find any reason why the values were missing.
The assumption were:**
1. Those people are not working
2. Technical issue with the form
3. They didn’t know / want to share this information.

**fixing the data:**
1. The column days_employed has negative values - we replaced the values with absolute ones.
2. The column education has inconsistent values, there are lower and upper case letters - we changed all to a lower case.
3. The column children has negative number (-1) and an outlier (20) - we replaced those values with 1 and 2 children. 
4. The column total_income has some missing values - we filled those rows with the median income according to the income type.
5. The dob_years column had some ‘0’ value - we replaced it with the median age.
6. The gender column had an unknown value XNA, we replaced it with male because the income was closer to a male income.

We had 72 duplicates and we dropped them and reset the index.

We add 2 categories to the data :age group and clean purpose.



# We checked the Hypotheses:
**The correlation between having children and paying back on time:**
- All people with 5 children paid on time.
- People with 0-4 children has 7.5-9.7% probability to not paying on time.
- People with no kids tends to ask for a loan more than others.


**The correlation between family status and paying back on time:**
- married people tends to ask for a loan more than others.
- All family status has 6.5-9.7% probability to not paying on time.
- unmarried and civil partnership tends to have more debts than others.

**The correlation between income level and paying back on time:**
- All income levels has 5.6-8.2% probability to not paying on time.
- People with total income of 10K-50K tends to ask more for a loan than others, 8.2% of them didn't pay on time.
- Only two people earn more than 300K and one of them didn't return the loan on time.

**The correlation between purpose and paying back on time:**
- most of the people are asking for a loan in order to buy a house and relatively they tend to pay back on time more than the others.
- 9% of people who ask a loan to buy a car or pay for education tends not to pay on time.
- Asking for a loan to a wedding is less common than others and 8% of them didn't pay on time.


<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
Good conclusion 👍
</div>

<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<h2> Overall conclusion <a class="tocSkip"></h2>
 
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
    
You've done a really good job! You've shown strong analytical and coding skills. You are already good at pandas, it's great! The preprocessing was done successfully, the rate for each group was found correctly. 
    
</div>    
    
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
    
There are some issues that need to be fixed, but it will not take much time. As soon as you fix them I will accept the project 😉 
    
</div>
    
    
<hr>    
    
    
Best regards,
    
    
Sveta
</div>